cpp/include/cudf/strings/detail/gather.cuh

```diff
@@ -34,6 +35,146 @@ namespace cudf {
+
+  auto const alignment_offset = reinterpret_cast<std::uintptr_t>(out_chars) % out_datatype_size;
+  uint4* out_chars_aligned    = reinterpret_cast<uint4*>(out_chars - alignment_offset);
+
+  for (size_type istring = global_warp_id; istring < total_out_strings; istring += nwarps) {
+    auto const out_start = out_offsets[istring];
+    auto const out_end   = out_offsets[istring + 1];
+
+    // This check is necessary because string_indices[istring] may be out of bound.
+    if (out_start == out_end) continue;
+
+    const char* in_start = strings_begin[string_indices[istring]].data();
+
+    // Both `out_start_aligned` and `out_end_aligned` are indices into `out_chars`.
+    // `out_start_aligned` is the first 16B aligned memory location after `out_start + 4`.
+    // `out_end_aligned` is the last 16B aligned memory location before `out_end - 4`. Characters
+    // between `[out_start_aligned, out_end_aligned)` will be copied using uint4.
+    // `out_start + 4` and `out_end - 4` are used instead of `out_start` and `out_end` to avoid
+    // `load_uint4` reading beyond string boundaries.
+    int32_t out_start_aligned =
+      (out_start + in_datatype_size + alignment_offset + out_datatype_size - 1) /
+        out_datatype_size * out_datatype_size -
+      alignment_offset;
+    int32_t out_end_aligned =
+      (out_end - in_datatype_size + alignment_offset) / out_datatype_size * out_datatype_size -
+      alignment_offset;
+
+    for (size_type ichar = out_start_aligned + warp_lane * out_datatype_size;
+         ichar < out_end_aligned;
+         ichar += cudf::detail::warp_size * out_datatype_size) {
+      *(out_chars_aligned + (ichar + alignment_offset) / out_datatype_size) =
+        load_uint4(in_start + ichar - out_start);
+    }
+
+    // Tail logic: copy characters of the current string outside `[out_start_aligned,
+    // out_end_aligned)`.
+    if (out_end_aligned <= out_start_aligned) {
+      // In this case, `[out_start_aligned, out_end_aligned)` is an empty set, and we copy the
+      // entire string.
+      for (int32_t ichar = out_start + warp_lane; ichar < out_end;
+           ichar += cudf::detail::warp_size) {
+        out_chars[ichar] = in_start[ichar - out_start];
+      }
+    } else {
+      // Copy characters in range `[out_start, out_start_aligned)`.
+      if (out_start + warp_lane < out_start_aligned) {
+        out_chars[out_start + warp_lane] = in_start[warp_lane];
+      }
+      // Copy characters in range `[out_end_aligned, out_end)`.
+      int32_t ichar = out_end_aligned + warp_lane;
+      if (ichar < out_end) { out_chars[ichar] = in_start[ichar - out_start]; }
+    }
+  }
+}
+
+// Strategy 2: Char-parallel
+// This strategy assigns characters to threads, and uses binary search for getting the string
+// index. To improve the binary search performance, fixed number of strings per threadblock is
+// used. This strategy is best suited for small strings.
+
```

Comment:
Can you convert this comment (and the one above) to doxygen style?

Reply:
Is this what you had in mind https://github.com/rapidsai/cudf/pull/7980/commits/e87da49f5b24310dd6dd40fd041afaffec94ddf5?

Reply:
Yes. That looks great.